lib/JMAP/Tester/UA/LWP.pm

```diff
@@ -0,0 +1,78 @@
+use v5.10.0;
+use warnings;
+
+package JMAP::Tester::UA::LWP;
+
+use Moo;
+with 'JMAP::Tester::Role::UA';
+
+use Future;
+
+has lwp => (
+  is   => 'ro',
+  lazy => 1,
+  default => sub {
+    my ($self) = @_;
+
+    require LWP::UserAgent;
+    my $lwp = LWP::UserAgent->new;
+    $lwp->cookie_jar({});
+
+    if ($ENV{IGNORE_INVALID_CERT}) {
+      $lwp->ssl_opts(SSL_verify_mode => 0, verify_hostname => 0);
+    }
+
+    return $lwp;
+  },
+);
+
+sub set_cookie {
+  my ($self, $arg) = @_;
+
+  for (qw(api_uri name value)) {
+    Carp::confess("can't set_cookie without $_") unless $arg->{$_};
+  }
+
+  my $uri = URI->new($arg->{api_uri});
+
+  $self->lwp->cookie_jar->set_cookie(
+    1,
+    $arg->{name},
+    $arg->{value},
+    '/',
+    $arg->{domain} // $uri->host,
+    $uri->port,
+    0,
+    ($uri->port == 443 ? 1 : 0),
+    86400,
+    0,
+    $arg->{rest} || {},
+  );
+}
+
+sub request {
+  my ($self, $tester, $req, $log_type, $log_extra) = @_;
```

Comment:
I wonder: would it be better to stick, in Role::UA:

```perl
has tester => (
  is => 'ro',
  weak_ref => 1,
);
```

because saying `$self->ua->request($self, $request, ...)` feels very strange to me.

I guess you probably didn't do this so that you can pass some extra agent in from afar, so you can say 

```perl
my $http = My::Cool::HTTP->new;
my $ua = JMAP::Tester::UA::Async->new({ client => $client });
my $tester = JMAP::Tester->new({ ua => $ua });
```

So maybe it's ok, but I just wanted to mention that it feels weird to me.

Reply:
Update: I have glanced at a couple of places that have to cope with the new code, and now I'm even more in favor of sticking a weak ref in, because `$tester->ua->request($tester, $req, 'foo')` feels dirty.

Reply:
Can you elaborate on "couple of places"?  I think they were all "code using Tester has to dig into the UA to make a misc. HTTP call", which is now fixed by providing http_request and friends.  I don't mind setting up the weak ref, but I'd _rather_ not if we can agree that this makes it unneeded.